Legacy support for old packet structures

JarvisCraft opened the pull request:
Feature proposed in #59 whose main point is to allow PacketWrapper to be used on lower versions in which structures of packet objects vary from these now.
This is reached by adding conditional statements which check minor version of minecraft server (this is stored as a static constant of `AbstractPacket` as an `int` primitive so the overhead of such approach seems to be close to zero).
Another enhancement being part of this PR is storing `ProtocolManager` instance from `ProtocolLibrary.getProtocolManager()` as a static constant of `AbstractPacket` which both guarantees no mismatches and performs a sort of microoptimization.

JarvisCraft committed:
Use static sonstant for ProtocolManager reference

JarvisCraft committed:
Fix access restrictions of PROTOCOL_MANAGER field

JarvisCraft committed:
Add version static constants

JarvisCraft committed:
Add ConversionUtil

JarvisCraft committed:
Add legacy support to WrapperPlayServerEntityTeleport

JarvisCraft commented:
@dmulloy2, waiting for your point on whether or not this PR should be continued.

JarvisCraft committed:
Inline local PROTOCOL_MANAGER referencing variable

JarvisCraft committed:
Remove unneeded import statement

JarvisCraft committed:
Add legacy support to WrapperPlayServerSpawnEntityLiving

JarvisCraft commented:
Another thing to discuss is what to do with those methods which don't have equivalents on older versions such as `setUniqueId(UUID)`.
Should these throw fail-fast exceptions on those versions or return some stubs instead?

JarvisCraft committed:
Add legacy support to WrapperPlayServerAttachEntity

dmulloy2 commented:
I like it. As for the methods without equivalents it should throw an UnsupportedOperationException

JarvisCraft committed:
Add UnsupportedOperationException for UUID methods in WrapperPlayServerSpawnEntityLiving

JarvisCraft committed:
Add UnsupportedOperationException for UUID methods in WrapperPlayServerSpawnEntityLiving

JarvisCraft committed:
Add legacy support to WrapperPlayServerSpawnEntity

JarvisCraft committed:
Merge remote-tracking branch 'origin/legacy-support' into legacy-support

JarvisCraft commented:
@dmulloy2, it seems that `WrapperPlayServerRelEntityMove` and `WrapperPlayServerRelEntityMoveLook` have different types for `setDx/y/z` and `getDx/y/z` as the first one uses raw `int` value which should be manually converted by the user and the second one uses value converted to `double` using `/ 4096D` so I will change the method signature in `EntityMove` so it follows the standard.

JarvisCraft commented:
Also will extend packets as in NMS so that `Entity` is parent of `RelEntityMove` `RelEntityMoveLook` and `EntityLook`.

JarvisCraft committed:
Add legacy support to Entity Move/Look packets

JarvisCraft committed:
Don't use wraps for TYPE constant

JarvisCraft committed:
Add legacy support to WrapperPlayServerEntityEquipment

JarvisCraft commented:
@dmulloy2, could you please review 7dd117acb10e135e966c2c591312058faca0d585

JarvisCraft committed:
Add minimal legacy support to WrapperPlayServerNamedSoundEffect

JarvisCraft committed:
Add legacy support to WrapperPlayServerNamedSoundEffect

JarvisCraft committed:
Remove irrelevant comments in WrapperPlayServerNamedSoundEffect

JarvisCraft committed:
Add legacy support to both WrapperPlayServer*SoundEffect

dmulloy2 commented:
@JarvisCraft sorry for the late reply on this, it's been a busy year.

What's the status of this PR? I looked over the diff and it looks good, but it is still a draft PR.

JarvisCraft commented:
@dmulloy2, hi there
I will try to finish it soon (had a big delay due to university and other stuff)
<3

JarvisCraft committed:
Merge pull request #1 from dmulloy2/master

Merge new commits from original repo

JarvisCraft committed:
Merge pull request #2 from JarvisCraft/master

Merge latest changes from origin

JarvisCraft committed:
Use non-snapshot version of ProtocolLib

Use latest non-snapshot version of ProtocolLib
This fixes build issues due to unfound dependency (4.4.0-SNAPSHOT seems to missing in at repo.dmulloy2.net)
Data taken from: http://repo.dmulloy2.net/nexus/service/rest/repository/browse/public/com/comphenix/protocol/ProtocolLib/

JarvisCraft committed:
Add comment about static final fields and major version check

JarvisCraft committed:
SPecify latest (Snapshot) version of ProtocolLib

JarvisCraft committed:
Merge pull request #3 from JarvisCraft/patch-1

Actualize version

JarvisCraft committed:
Use release version of ProtocolLin

JarvisCraft committed:
Fix broken packet names

JarvisCraft committed:
Merge pull request #4 from dmulloy2/master

Sync with origin

JarvisCraft committed:
Merge branch 'legacy-support' into legacy-support-upstream-merge

JarvisCraft committed:
Merge pull request #5 from JarvisCraft/legacy-support-upstream-merge

Actualize fork

JarvisCraft committed:
Cleanup text related to legacy-support in AbstractPacket

JarvisCraft committed:
Make AbstractPacket.handle final

JarvisCraft committed:
Delegate logic of misspelled deprecate entries to correct ones

JarvisCraft committed:
Add BackwardsCompatible and BackwardsCompatibility

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientAbilities

JarvisCraft committed:
Make Backwards Compatibility annotations easier to write

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientArmAnimation

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientAutoRecipe

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientBeacon

JarvisCraft committed:
Specify Minecraft 1.8 as minimal backwards compatible

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientBlockDig

JarvisCraft committed:
Replace javadocs-like copyrights with correct ones

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientBlockPlace

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientBoatMove

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientBookEdit

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientChat

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientClientCommand

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientCloseWindow

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientCustomPayload

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientDifficultyChange

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientDifficultyLock

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientEnchantItem

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientEntityAction

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientEntityNbtQuery

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientFlying

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientHeldItemSlot

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientItemName

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientKeepAlive

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientLook

This also updates WrapperPlayClientFlying

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientPickItem

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientPosition

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientPositionLook

JarvisCraft committed:
BackwardsCompat: WrapperPlayServerEntity

JarvisCraft committed:
BackwardsCompat: WrapperPlayServerRelEntityMove

JarvisCraft committed:
BackwardsCompat: WrapperPlayServerRelEntityLook

JarvisCraft committed:
BackwardsCompat: WrapperPlayServerRelEntityMoveLook

JarvisCraft committed:
chore: use more readable version flags

JarvisCraft committed:
Fix copyright of ItemFactoryDelegate

JarvisCraft committed:
Rework WrapperPlayClientFlying and its relatives to correspond to packets

JarvisCraft committed:
Add TODOs to WrapperPlayClientRecipeDisplayed

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientResourcePackStatus

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientSetCommandBlock

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientSetCommandMinecart

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientSetCreativeSlot

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientSetJigsaw

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientSettings

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientSpectate

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientSteerVehicle

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientStruct

JarvisCraft committed:
Remove WrapperPlayClientStructureBlock duplicating WrapperPlayClientStruct

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientTabComplete

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientTeleportAccept

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientTileNbtQuery

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientTradeSelect

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientTransaction

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientUpdateSign

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientUseEntity

JarvisCraft committed:
Resolve issues with backwards compatibility of WrapperPlayClientArmAnimation

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientUseItem

This is exact class provides its specific mappings between versions
which should be moved to ProtocolLib

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientVehicleMove

JarvisCraft committed:
BackwardsCompat: WrapperPlayClientWindowClick

JarvisCraft committed:
Cleanup TestWrappers

JarvisCraft commented:
Welp, I'm weird thus it took me another year to sit and start doing it intensively.

Now all `WrapperPlayClient*` part is backwards compatible.

Significant details:
- I've added `@BackwardsCompatible` annotation for easier tracking of backwards compatibility
- `WrapperPlayClientUseItem` contains a compound structure `MovingObjectPosition` which should probably be implemented in ProtocolLib itself
- Tests won't run now (RIP) as `AbstractPacket` depends on `ProtocolLibrary.getProtocolManager()` which is `null` when running tests

The next stage is doing the same job for `WrapperPlayServer*`

JarvisCraft committed:
Reformat all WrapperPlayClient* packets

JarvisCraft commented:
Also, there was code-style mess as some classes were using tabs and some were using spaces thus I've used tabs everywhere (as it was more common among default classes) thus if spaces are the preferred indents (I personally prefer them) it won't be a problem to use them instead.

JarvisCraft commented:
As another point for discussion, there are some legacy details which may be also influenced if you support it:
- I suggest removing legacy methods with typos which were replaced with correct ones. As PacketWrapper is mostly used as a shaded dependency it should not cause much problems on existing servers.
- There are some incorrect method names such as `getEntityID` instead of `getEntityId` which stand against Java's code conventions and mess with methods like `getKeepAliveId` thus I also suggest removing them
- Some `boolean` accessors use `is` prefix like `isInvulnerable()` while others use `get` and some simply omit these, I think that it's worth using `get` everywhere as, while I prefer `is` for `boolean`s, here this are not just getters bu reflective accessors

JarvisCraft committed:
BackwardsCompat: WrapperPlayServerAbilities

JarvisCraft committed:
build(dep): add org.jetbrains:annotations

JarvisCraft commented:
Another big job to be done is Advancements as they seem to be broken as I remember.
I will probably do the same job as done with `MovingObjectPosition`: implement wrappers in PacketWrapper leaving an opportunity for you to move it into ProtocolLib.

JarvisCraft committed:
test: temporary disable tests as they are currently incompatible with BackwardsCompatibility support

JarvisCraft committed:
build(dep): update ProtocolLib to 4.6.0-SNAPSHOT

JarvisCraft committed:
BackwardsCompat: WrapperPlayServerEntityEquipment

JarvisCraft committed:
BackwardsCompat: WrapperPlayServerSpawnEntityExperienceOrb

JarvisCraft committed:
BackwardsCompat: WrapperPlayServerSpawnEntity

JarvisCraft committed:
feat: add VersionUtil and EntityUtil

JarvisCraft committed:
fix: replace incorrect usages of MAJOR_VERSION instead of MAJOR_VERSION

JarvisCraft committed:
fix: use EntityUtil in WrapperPlayServerSpawnEntity

JarvisCraft committed:
chore: initialize PROTOCOL_MANAGER lazily

JarvisCraft committed:
BackwardsCompat: WrapperPlayServerSpawnEntityLiving

JarvisCraft committed:
chore: bump version to 1.16.4-R0.1-SNAPSHOT

JarvisCraft committed:
fix: use correct Regex groups in VersionUtil

JarvisCraft committed:
fix: replace NMS imports in tests with the correct ones

JarvisCraft committed:
fix: resolve nullability issues in EntityUtil

JarvisCraft committed:
chore: respect nulls in WrapperPlayServerEntityEquipment

JarvisCraft committed:
build: temporarily use personal GitHub Package Registry as a Maven repository

JarvisCraft committed:
build: temporarily use personal GitHub Package Registry as a Maven repository

JarvisCraft commented:
I have temporarily configured the branch of my fork to deploy artifacts to GH Package Registry so that it is possible to depend on this patch while it is not implemented.
Once it becomes non-draft I will tag you and reset 330d078

JarvisCraft commented:
@dmulloy2 could you please allow me to temporarily deploy current legacy-support builds under my coordinates (i.e. `ru.progrm-jarvis.minecraft:PacketWrapper`) at Sonatype OSSRH so that in can be used as a dependency now until the patch is fully ready to be meged into upstream.

At current it gets deployed to [GitHub Package Registry](https://github.com/JarvisCraft/PacketWrapper/packages/586527) for the same purpose but it requires user to provide access credentials (even though the repo is public) which seems to be an issue:
for example I have [minecraft-utils](https://github.com/JarvisCraft/minecraft-util) depending on this, then [custom-stuff](https://github.com/divine-craft/custom-stuff) depends on `minecraft-utils` and than multiple projects depend on `custom-stuff` and yet each of them has to provide GitHub credentials to be able to compile just because the top-level dependency (`minecraft-utils` uses GitHub Package Registry's PacketWrapper).

dmulloy2 commented:
@JarvisCraft yeah you're absolutely free to deploy your own version under your own coordinates

JarvisCraft commented:
> @JarvisCraft yeah you're absolutely free to deploy your own version under your own coordinates

Thanks! Now it is available via [Sonatype OSSRH](https://oss.sonatype.org/content/repositories/snapshots/ru/progrm-jarvis/minecraft/packet-wrapper/). Once the job on this PR is ready (once it will happen, haha) I will also open a PR to enable deployment under your coordinates to Sonatye OSSRH and possibly Maven Central.

JarvisCraft committed:
Merge remote-tracking branch 'origin/legacy-support' into legacy-support

# Conflicts:
#	.github/workflows/deploy.yml
#	PacketWrapper/pom.xml

JarvisCraft committed:
feat: implement AbstractPacket#toString()

JarvisCraft committed:
ci: remove redundant settings file

dmulloy2 commented:
Also you should be good to go back to using my maven repo

dmulloy2 commented:
@JarvisCraft any update on this?

BlcDragon committed:
fix: do not write both shorts and integers

JarvisCraft committed:
Merge pull request #7 from BlcDragon/patch-2

fix: do not write both shorts and integers

BlcDragon committed:
build: bump ProtocolLib version to 4.7.0

JarvisCraft committed:
Merge pull request #8 from BlcDragon/legacy-support

build: bump ProtocolLib version to 4.7.0
